### ISSUE TITLE
speed up the process: options

### ISSUE THREAD

**Issue by vajnorcan** [solowt/Client-side-viewshed#3]
thanks for the tool, helped me lot to resolve the viewshed requirement for my application. Even though it's not entirely slow, I'm looking at the options of speeding it up. 

You have mentioned that

> You could also attempt to "stitch" together different rasters and do the elevation lookups directly on the resulting raster, instead of having to search through the entire tile tree.

I wonder how would you do the lookup then if it's the raster (is it a raster after all??) ? Map algebra?

Thanks

**Comment by solowt** [comment]
Glad to hear it's been helpful!

What I meant by that comment is that most of the computation time is spent in [this loop](https://github.com/solowt/Client-side-viewshed/blob/master/ClientVS.js#L71), which creates an array of all elevation values needed for the viewshed by querying the view.  The views get its elevation data from an image service ([this](https://elevation3d.arcgis.com/arcgis/rest/services/WorldElevation3D/Terrain3D/ImageServer) is the default public one).

I'll take a look at how you might speed this up.  One option would be using an [elevation sampler](https://developers.arcgis.com/javascript/latest/api-reference/esri-layers-support-ElevationSampler.html) which was added to the Esri JS API in the past year.

**Comment by vajnorcan** [comment]
I've rewritten the code to elevationSampler as I'm using 4.8, but still have asked the question, as it's UI blocking operation (quite heavy computing there). Hence trying to rewrite the app (Angular6 + ESRI JS API 4.8) to use the webworkers but it's the pain in the ... due to lots of async dependencies.

Update: there is a new weboworkers interface that simplify the work with the dependencies. I'll look into it
https://developers.arcgis.com/javascript/latest/api-reference/esri-core-workers.html

**Comment by solowt** [comment]
I updated the master branch to use `elevationSampler`.  I notice it's faster if I query the sampler with a single multipoint containing all points that need to the sampled.

In terms of moving the whole thing to another thread, I think this is definitely feasible.  One approach would be transferring the sampler to the other thread.  This is possible, but would require some work.  A simpler idea would be creating a new elevation layer in the other thread and then creating the sampler there.

I'll take a look at that option.

**Comment by vajnorcan** [comment]
Well, AFAIK you can transfer only simple object, I don't think it would handle the whole sampler. Not sure if ground can be stringified, but seems like just getting the url of the elevation layer, and querying it (or using the sampler on it) from within the webworker while passing the 'raster' could work ..
I will give it few more hours if not, I'll move the whole app into webworker (Angular6)

Update: well I've managed to get the elevationSampler from the layer that I've created in the webworker. The downside is, that creating a webworker via esri's webworker takes a while as it nearly loads the whole js api into the worker and that process takes a while (network tab goes crazy!) and it kinds of beats the purpose. This happens only once, and also, the variables are persisted which might be alright. Having said that, I don't think this exercise was worth anything, perhaps only that webworkers are hard to combine with the esri js api. Service workers however, could be the right beast.

FYI I wasn only requesting the following modules 
```
        'esri/core/promiseUtils',
        'esri/layers/ElevationLayer',
        'esri/geometry/Extent',
        'dojo/Deferred',
```
and that resulted in 521 requests that took 18sec to load.

**Comment by solowt** [comment]
Interesting.  I switched the whole thing to be worker compatible on a new branch `workers`.  It seems to work okay for me, you can take a look at the gh demo: https://solowt.github.io/Client-side-viewshed/ or that branch (see `index-workers.js` for the code that loads the worker).

Are you passing in a strategy to `workers.open`?  Looks like the default is to load the code in all workers, but if I pass in `{ stategy: "dedicated" }` as the second argument to `workers.open`, it will only load the file in one thread.

**Comment by vajnorcan** [comment]
That makes sense now, as it had to load as many times as there are cores (in my case x4). I've taken your sample and moved the worker connection generating into the 'click' event and I've watched the number of requests of the map load... then cleared the console and then clicked the map. It requested the  144 requests for your worker, which would be kind of acceptable, but I think more acceptable would be the following idea here from Rene
https://community.esri.com/thread/219307-web-worker-and-js-api-modules

The idea is to build your own custom build for the worker. I think I'll give it a go in a separate project and each time I need to get the new worker spinned off, I'll just add more modules in that project build it, and then include it in my project as and asset to be loaded.

**Comment by solowt** [comment]
Nice, creating a custom build sounds like a good approach.

**Comment by solowt** [comment]
I had a chance to make another optimization - the calculation should be much faster in situations where the resulting polygon has many rings.

Closing this issue
